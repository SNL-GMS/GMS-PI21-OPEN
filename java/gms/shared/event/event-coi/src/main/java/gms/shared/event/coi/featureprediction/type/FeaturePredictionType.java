package gms.shared.event.coi.featureprediction.type;

import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import gms.shared.event.coi.featureprediction.value.ArrivalTimeFeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.FeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.NumericFeaturePredictionValue;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;

/**
 * Class for an instantiable object that represents the type of feature prediction. Each "type"
 * is associated with a FeaturePredictionValue class; together, they fully define what each type of
 * feature prediction looks like.
 *
 * @param <T> The FeaturePredictionValue class that should be coupled with this type.
 */
@JsonDeserialize(using = FeaturePredictionType.Deserializer.class)
public abstract class FeaturePredictionType<T extends FeaturePredictionValue<?, ?, ?>> {

  public enum TypeNames {
    ARRIVAL_TIME,
    SLOWNESS,
    EMERGENCE_ANGLE,
    RECEIVER_TO_SOURCE_AZIMUTH,
    SOURCE_TO_RECEIVER_AZIMUTH,
    SOURCE_TO_RECEIVER_DISTANCE,
  }

  private final TypeNames name;

  private FeaturePredictionType(
    TypeNames name
  ) {
    this.name = name;
  }

  @JsonValue
  public TypeNames getName() {
    return name;
  }

  @Override
  public String toString() {
    return "FeaturePredictionType{" +
      "name='" + name + '\'' +
      '}';
  }

  public abstract Class<T> getTypeValueClass();

  //
  // Constants for the supported feature prediction types.
  //

  public static final FeaturePredictionType<ArrivalTimeFeaturePredictionValue> ARRIVAL_TIME_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.ARRIVAL_TIME) {
      @Override
      public Class<ArrivalTimeFeaturePredictionValue> getTypeValueClass() {
        return ArrivalTimeFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SLOWNESS_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SLOWNESS) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> EMERGENCE_ANGLE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.EMERGENCE_ANGLE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.RECEIVER_TO_SOURCE_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_DISTANCE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  /**
   * Custom deserializer used by Jackson. The serialized JSON should contain a single field with
   * a single string that is the name of one of the types, and that string will be returned
   * be JsonParser::getValueAsString.
   */
  static class Deserializer extends StdDeserializer<FeaturePredictionType<?>> {

    private static final Map<TypeNames, FeaturePredictionType<?>> nameTypeMap = Map.of(
      TypeNames.ARRIVAL_TIME, ARRIVAL_TIME_PREDICTION_TYPE,
      TypeNames.SLOWNESS, SLOWNESS_PREDICTION_TYPE,
      TypeNames.EMERGENCE_ANGLE, EMERGENCE_ANGLE_PREDICTION_TYPE,
      TypeNames.RECEIVER_TO_SOURCE_AZIMUTH, RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_AZIMUTH, SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_DISTANCE, SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE
    );

    public Deserializer() {
      this(null);
    }

    public Deserializer(Class<?> vc) {
      super(vc);
    }

    @Override
    public FeaturePredictionType<?> deserialize(JsonParser p, DeserializationContext ctxt)
      throws IOException {
      return Optional.ofNullable(nameTypeMap.get(TypeNames.valueOf(p.getValueAsString())))
        .orElseThrow();
    }
  }
}
