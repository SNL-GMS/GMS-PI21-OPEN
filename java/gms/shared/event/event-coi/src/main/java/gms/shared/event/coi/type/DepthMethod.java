package gms.shared.event.coi.type;

import com.google.common.base.Functions;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;

import java.util.Arrays;

public enum DepthMethod {
  A("a"),
  F("f"),
  D("d"),
  R("r"),
  G("g"),
  UNKNOWN("-");


  private static final ImmutableMap<String, DepthMethod> LABELS_BY_MODEL;

  static {
    LABELS_BY_MODEL = Arrays.stream(values())
      .collect(ImmutableMap.toImmutableMap(DepthMethod::getLabel, Functions.identity()));
  }

  public final String label;


  DepthMethod(String label) {
    this.label = label;
  }

  public static DepthMethod valueOfLabel(String label) {
    Preconditions.checkNotNull(label, "label cannot be null");
    Preconditions.checkArgument(!label.isBlank(), "label cannot be blank");
    return LABELS_BY_MODEL.get(label);
  }

  public String getLabel() {
    return label;
  }

}

