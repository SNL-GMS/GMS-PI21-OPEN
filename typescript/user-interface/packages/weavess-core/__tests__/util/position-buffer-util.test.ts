/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable @typescript-eslint/no-floating-promises */

import type { DataBySampleRate } from '../../src/ts/types';
import type {
  CreatePositionBufferBySampleRateParams,
  CreatePositionBufferByTimeParams
} from '../../src/ts/util/position-buffer-util';
import {
  convertToPositionBuffer,
  createPositionBufferForDataBySampleRate,
  createPositionBufferForDataByTime,
  getBoundsForPositionBuffer
} from '../../src/ts/util/position-buffer-util';
import type { WeavessTypes } from '../../src/ts/weavess-core';

describe('WEAVESS Core: Position Buffer Utils', () => {
  test('functions should be defined', () => {
    expect(createPositionBufferForDataByTime).toBeDefined();
    expect(createPositionBufferForDataBySampleRate).toBeDefined();
  });

  test('createPositionBufferForDataByTime', () => {
    const data: CreatePositionBufferByTimeParams = {
      displayStartTimeSecs: 0,
      displayEndTimeSecs: 10,
      glMin: 0,
      glMax: 100,
      values: [
        {
          timeSecs: 0,
          value: 5
        },
        {
          timeSecs: 5,
          value: 8
        },
        {
          timeSecs: 9,
          value: 2
        }
      ]
    };
    const buffer = createPositionBufferForDataByTime(data);
    expect(buffer[0]).toEqual(0);
    expect(buffer[1]).toEqual(5);
    expect(buffer[2]).toEqual(50);
    expect(buffer[3]).toEqual(8);
    expect(buffer[4]).toEqual(90);
    expect(buffer[5]).toEqual(2);
  });

  test('createPositionBufferForDataBySampleRate', () => {
    const data: CreatePositionBufferBySampleRateParams = {
      startTimeSecs: 0,
      endTimeSecs: 100,
      sampleRate: 5,
      displayStartTimeSecs: 0,
      displayEndTimeSecs: 10,
      glMin: 0,
      glMax: 100,
      values: [4, 5, 7]
    };
    const buffer = createPositionBufferForDataBySampleRate(data);
    expect(buffer[0]).toEqual(0);
    expect(buffer[1]).toEqual(4);
    expect(buffer[2]).toEqual(2);
    expect(buffer[3]).toEqual(5);
    expect(buffer[4]).toEqual(4);
    expect(buffer[5]).toEqual(7);
  });

  const testArray = new Array(100).fill(0, 0, 100).map((val: number, index: number) => index);

  const sampleWave: DataBySampleRate = {
    /** Epoch start time in seconds */
    startTimeSecs: 0,
    endTimeSecs: 100,
    sampleRate: 40,
    values: testArray
  };

  const visibleDomain: WeavessTypes.TimeRange = {
    startTimeSecs: 0,
    endTimeSecs: 1000
  };

  it('can convert an array of numbers into typed array', () => {
    const convertedWave = convertToPositionBuffer(sampleWave, visibleDomain);
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    expect(convertedWave.BYTES_PER_ELEMENT).toBe(4);
  });

  it('creates a TypedArray with the correct length', () => {
    const convertedWave = convertToPositionBuffer(sampleWave, visibleDomain);
    expect(convertedWave).toHaveLength(sampleWave.values.length * 2);
  });

  it('can convert an array of numbers into an equivalent typed array', () => {
    const convertedWave = convertToPositionBuffer(sampleWave, visibleDomain);
    const yValues = convertedWave.filter((v, i) => i % 2 === 1);
    expect(Array.from(yValues)).toMatchObject(Array.from(sampleWave.values));
  });

  it('gives an empty TypedArray when given an empty set of values', () => {
    const emptyWave = {
      ...sampleWave,
      values: []
    };
    const convertedWave = convertToPositionBuffer(emptyWave, visibleDomain);
    expect(convertedWave).toHaveLength(0);
  });

  it('throws when values is undefined', () => {
    const wipeoutWave = {
      ...sampleWave,
      values: undefined
    };
    expect(() =>
      convertToPositionBuffer(wipeoutWave.values, visibleDomain)
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws when waveform is undefined', () => {
    expect(() => convertToPositionBuffer(undefined, visibleDomain)).toThrowErrorMatchingSnapshot();
  });

  it('throws when visibleDomain is undefined', () => {
    expect(() => convertToPositionBuffer(sampleWave, undefined)).toThrowErrorMatchingSnapshot();
  });

  describe('getBoundsForPositionBuffer', () => {
    //                             index =>    0   1   2  3   4    5  6  7   8   9
    const positionBuffer = Float32Array.from([-10, 0, -5, 100, 0, -5, 5, 10, 10, -5]);

    it('exists', () => {
      expect(getBoundsForPositionBuffer).toBeDefined();
    });

    it('throws when it is given an odd number of elements', () => {
      expect(() => getBoundsForPositionBuffer(Float32Array.from([0, 0, 1, 100, 2]))).toThrow();
    });

    it('finds the min and max with no start or end indices', () => {
      const { min, max } = getBoundsForPositionBuffer(positionBuffer);
      expect(min).toBe(-5);
      expect(max).toBe(100);
    });

    it('finds the min and max with given start index', () => {
      const { min, max } = getBoundsForPositionBuffer(positionBuffer, 3);
      expect(min).toBe(-5);
      expect(max).toBe(100);
    });

    it('finds the min and max with given start and end index', () => {
      const { min, max } = getBoundsForPositionBuffer(positionBuffer, 3, 5);
      expect(min).toBe(-5);
      expect(max).toBe(100);
    });
  });
});
