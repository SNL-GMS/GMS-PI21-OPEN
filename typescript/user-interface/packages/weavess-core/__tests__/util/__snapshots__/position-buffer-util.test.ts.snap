// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WEAVESS Core: Position Buffer Utils throws when values is undefined 1`] = `"Typed array conversion failed; Invalid waveform should contain an array of numbers."`;

exports[`WEAVESS Core: Position Buffer Utils throws when visibleDomain is undefined 1`] = `"Typed array conversion failed; No visible domain was provided."`;

exports[`WEAVESS Core: Position Buffer Utils throws when waveform is undefined 1`] = `"Typed array conversion failed; Invalid waveform should contain an array of numbers."`;
